botcore/db.py

```diff
@@ -0,0 +1,162 @@
+    CERTIFICATE_FILE = "config/firebase_credentials.json"
+    COL_MEMBERS = "members"
+    COL_SECRETS = "secrets"
+
+    def __init__(self, bot):
+        self.bot = bot
+        self.db = None
+        self._connect()
+
+    def get_member_data(self, id):
+        """
+        Retrieve entry for a member from the database.
+
+        Args:
+            id: Discord ID of a member.
+        
+        Returns:
+            Dictionary containing all keys and values associated with a member.
+        
+        Raises:
+            MemberNotFound: If member does not exist in the database.
+        """
+
+        data = self._get_member_doc(id).get().to_dict()
+        if data is None:
+            raise MemberNotFound
+        return data
+
+    def get_unverified_members_data(self):
+        """
+        Retrieve entries for all unverified members from the database.
+
+        Returns:
+            Dictionary where each key is a member ID and each value is the info
+            associated with that member.
+        """
+
+        unverified = {}
+        docs = self._get_members_col() \
+            .where("id_verified", "==", False).stream()
+        for doc in docs:
+            member_id = int(doc.id)
+            member_data = doc.to_dict()
+            unverified[member_id] = member_data
+        return unverified
+
+    def set_member_data(self, id, info):
+        """
+        Write entry for a member to the database. If entry already exists, will
+        replace it.
+
+        Args:
+            id: Discord ID of a member.
+            info: Dictionary of keys and values to write.
+        """
+
+        self._get_member_doc(id).set(info)
+
+    def update_member_data(self, id, info, must_exist=True):
+        """
+        Update entry for a member in the database. Will only modify given keys
+        and values. If a key does not already exist, it will be created.
+
+        By default, will raise an exception if the member does not exist in the
+        database.
+
+        Args:
+            id: Discord ID of a member.
+            info: Dictionary of keys and values to write.
+            must_exist: If the member must exist in the database. If False,
+                will create a new entry if they do not.
+
+        Raises:
+            MemberNotFound: If the member does not exist in the database and
+                must_exist == False.
+        """
+
+        try:
+            self._get_member_doc(id).update(info)
+        except google.cloud.exceptions.NotFound:
+            raise MemberNotFound
+
+    def get_secret(self, id):
+        """
+        Retrieve entry for a secret from the database.
+        
+        Args:
+            id: ID of a secret.
+
+        Returns:
+            Secret bytes associated with id or None if no such secret exists.
+        """
+
+        doc = self._get_secrets_col().document(str(id)).get().to_dict()
+        if doc is None:
+            return None
+        return doc["secret"]
```

Comment:
A lot of your code assumes the schema in firebase is intact. What happens if someone breaks it via the UI? Not something you need to consider now, just a thought

Reply:
This is a good point. I'll make this an issue